admittance_controller/src/admittance_controller.cpp

```diff
@@ -0,0 +1,602 @@
+{
+controller_interface::CallbackReturn AdmittanceController::on_init()
+{
+  // initialize controller config
+  try
+  {
+    parameter_handler_ = std::make_shared<admittance_controller::ParamListener>(get_node());
+    admittance_ = std::make_unique<admittance_controller::AdmittanceRule>(parameter_handler_);
+  }
+  catch (const std::exception & e)
+  {
+    RCLCPP_ERROR(
+      get_node()->get_logger(), "Exception thrown during init stage with message: %s \n", e.what());
+    return controller_interface::CallbackReturn::ERROR;
+  }
+
+  // number of joints in controllers is fixed after initialization
+  num_joints_ = admittance_->parameters_.joints.size();
+
+  // allocate dynamic memory
+  last_reference_.positions.assign(num_joints_, 0.0);
+  last_reference_.velocities.assign(num_joints_, 0.0);
+  last_reference_.accelerations.assign(num_joints_, 0.0);
+  last_commanded_ = last_reference_;
+  reference_ = last_reference_;
+  reference_admittance_ = last_reference_;
+  joint_state_ = last_reference_;
+
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::InterfaceConfiguration AdmittanceController::command_interface_configuration()
+  const
+{
+  if (!admittance_)
+  {
+    return {controller_interface::interface_configuration_type::INDIVIDUAL, {}};
+  }
+
+  std::vector<std::string> command_interfaces_config_names;
+  for (const auto & interface : admittance_->parameters_.command_interfaces)
+  {
+    for (const auto & joint : command_joint_names_)
+    {
+      auto full_name = joint + "/" + interface;
+      command_interfaces_config_names.push_back(full_name);
+    }
+  }
+
+  return {
+    controller_interface::interface_configuration_type::INDIVIDUAL,
+    command_interfaces_config_names};
+}
+
+controller_interface::InterfaceConfiguration AdmittanceController::state_interface_configuration()
+  const
+{
+  if (!admittance_)
+  {
+    return {controller_interface::interface_configuration_type::INDIVIDUAL, {}};
+  }
+
+  std::vector<std::string> state_interfaces_config_names;
+  for (size_t i = 0; i < admittance_->parameters_.state_interfaces.size(); ++i)
+  {
+    const auto & interface = admittance_->parameters_.state_interfaces[i];
+    for (const auto & joint : admittance_->parameters_.joints)
+    {
+      auto full_name = joint + "/" + interface;
+      state_interfaces_config_names.push_back(full_name);
+    }
+  }
+
+  auto ft_interfaces = force_torque_sensor_->get_state_interface_names();
+  state_interfaces_config_names.insert(
+    state_interfaces_config_names.end(), ft_interfaces.begin(), ft_interfaces.end());
+
+  return {
+    controller_interface::interface_configuration_type::INDIVIDUAL, state_interfaces_config_names};
+}
+
+std::vector<hardware_interface::CommandInterface>
+AdmittanceController::on_export_reference_interfaces()
+{
+  // create CommandInterface interfaces that other controllers will be able to chain with
+  if (!admittance_)
+  {
+    return {};
+  }
```

Comment:
It shouldn't be, but its is sort of a safety measure. This could get called in test code for example because it follows separate logic from the controller manger.